owncloudData/src/main/java/com/owncloud/android/data/oauth/datasource/impl/RemoteOAuthDataSourceImpl.kt

```diff
@@ -60,4 +65,24 @@ class RemoteOAuthDataSourceImpl(
 
         return remoteTokenResponseMapper.toModel(tokenResponse)!!
     }
+
+    override fun registerClient(clientRegistrationRequest: ClientRegistrationRequest): ClientRegistrationInfo {
+        val ownCloudClient =
+            clientManager.getClientForUnExistingAccount(clientRegistrationRequest.registrationEndpoint, false)
+
+        val remoteClientRegistrationInfo = executeRemoteOperation {
+            oidcService.registerClientWithRegistrationEndpoint(
+                ownCloudClient = ownCloudClient,
+                clientRegistrationParams = ClientRegistrationParams(
+                    registrationEndpoint = clientRegistrationRequest.registrationEndpoint,
+                    clientName = clientRegistrationRequest.clientName,
+                    redirectUris = clientRegistrationRequest.redirectUris,
+                    tokenEndpointAuthMethod = clientRegistrationRequest.tokenEndpointAuthMethod,
+                    applicationType = clientRegistrationRequest.applicationType
+                )
+            )
+        }
+
+        return remoteClientRegistrationInfoMapper.toModel(remoteClientRegistrationInfo)!!
```

Comment:
!! 😮

Reply:
Mappers :( I don't like it either. `remoteClientRegistrationInfo` is not nullable, so `toModel` will return a non-nullable model. It is done that way since the beginning. I could create an issue to replace mappers (at least for remote) with extensions since they introduce `!!`

Reply:
Perfect